draft: add plugin support for KMSs (#1901)
sigstore/sigstore: pkg/signature/kms/cliplugin/README.md

# CLI Plugin

This is a package and module for using custom KMS plugins as separate executables.
It is intended to be used by cosign, but you may use this in your own programs that import sigstore.

## Design

We follow [kubectl's style](https://kubernetes.io/docs/tasks/extend-kubectl/kubectl-plugins/#writing-kubectl-plugins) of plugins. Any language that can create CLI programs can be used to make a plugin.

### Usage

Plugins are separate programs on your system's PATH, named in the scheme `sigstore-kms-[name]`, like `sigstore-kms-my-hsm`. They can be invoked with cosign like `cosign [sub-command] --key "my-hsm://my-key-id" ...
`

### Protocol

The main program will invoke the program with these specifications:

* stdin
  * Data to be signed or verified.
* arg 1

>   * A number idnetifying the version of this protocol.

Reviewer: ```suggestion
  * A number identifying the version of this protocol.
```

Author: fixed.

    * In the future we may change this protocol, either the encoding or the formatting of argument and return values.

>     These are necessarily breaking changes, and will be major-version bumps to sigstore and cosign.

Reviewer: ```suggestion
    Protocol changes will result in a major-version bump for the library.
```

Reviewer: Removing a mention of Cosign, as the maintainers may not want to do a major bump for this.

Author: fixed.

    When changing the protocol, new versions of sigstore will not maintain backwards compatibility with
    previous protocol versions. If a plugin author wishes, they may branch their plugin program’s behaviour
    to be compatible with multiple versions of the protocol, or multiple major versions of sigstore and cosign.
* arg 2
  * JSON of initialization options and method arguments.
* stdout
  * JSON of method return values.

See [./common/interface.go](./common/interface.go) for the full JSON schema.

The plugin program must first exit before sigstore begins parsing responses.

#### Error Handling

The plugin program’s stderr will be redirected to the main program’s stderr. This way, the main program may also see the plugin program’s debug messages.

Plugin authors may return errors with `PluginResp.ErrorMessage`, but the plugin's exit status will be ignored.

### Implementation

> Plugin authors must implement the `kms.SignerVerifier` interface methods in their chosen language. Each method will inbvoke your program once, and the reponse will be parsed from stdout.

Reviewer: ```suggestion
Plugin authors must implement the `kms.SignerVerifier` interface methods in their chosen language. Each method will invoke your program once, and the response will be parsed from stdout.
```

Author: fixed.

> Exit status is ignored. Your programs stderr will be redirected to the main program, and errors you wish to return must be serialized in `PluginResp.ErrorMessage` in stdout.

Reviewer: ```suggestion
Exit status is ignored. Your program's stderr will be redirected to the main program, and errors you wish to return must be serialized in `PluginResp.ErrorMessage` in stdout.
```

For authors using Go, we vend some helper functions to help you get started.

* `GetPluginArgs(osArgs []string) (*PluginArgs, error)`

>   * to parse the `PluginArgs`. The plugin author can use the values for some pre-processing, perhaps with the values in `PluginArgs.InitOptions`.

Reviewer: tiny nit, i'd make each bullet point one line rather than have single sub bullet points

* `Dispatch(stdout io.Writer, stdin io.Reader, pluginArgs *PluginArgs, impl kms.SignerVerifier) (*PluginResp, error)`
  * for auto-invoking the correct interface methods, given an implementation. stdin will be the message argument for the methods that sign and verify data. The plugin author may do post-processing with the returned PluginResp.
* `WriteResponse(wr io.Writer, resp *PluginResp) error`
  * , should the author wish to construct and return their own PluginResp and exit.
* `WriteErrorResponse(wr io.Writer, err error) error`
  * , should the author simply send an error message and exit.

## Development

Changes to the `SignerVerifier` interface are to be handled in [./signer.go's](./signer.go) `PluginClient` and [./handler/dispatch.go's](./handler/dispatch.go) `Dispatch()`.

### Adding New Methods or Method Options

> Adding new methods or options are *not* necessarily breaking changes to the schemas, so we may consider these to be minor version increments, both to the protocol version and the sigstore version.

Reviewer: When would adding a method be a breaking change?

Author: A new method (with new version) might break the plugin programs, if the plugin does not respect the new protocol version or the new method is not implemented by the plugin.

It's plausible that a new method might be rarely used, like `CreateKey()`, for example.

Added a TODO to implement semver, probably with https://pkg.go.dev/golang.org/x/mod/semver

# CLI Plugin

This is a package and module for using custom KMS plugins as separate executables. 
It is intended to be used by cosign, but you may use this in your own programs that import sigstore.

## Design

We follow [kubectl's style](https://kubernetes.io/docs/tasks/extend-kubectl/kubectl-plugins/#writing-kubectl-plugins) of plugins. Any language that can create CLI programs can be used to make a plugin.

### Usage

> Plugins are separate programs on your system's PATH, named in the scheme `sigstore-kms-[name]`, like `sigstore-kms-my-hsm`. They can be invoked with cosign like `cosign [sub-command] --key "my-hsm://my-key-id" ...

Author: For now, if there is a clash, it will run the one that appears first in PATH. It won't error due to a clash. We could allow the user to specify the location of the executable with an env variable.

`

### Protocol

The main program will invoke the program with these specifications:

* stdin
  * Data to be signed or verified.
* arg 1
  * A number idnetifying the version of this protocol.
    * In the future we may change this protocol, either the encoding or the formatting of argument and return values.
    These are necessarily breaking changes, and will be major-version bumps to sigstore and cosign.
    When changing the protocol, new versions of sigstore will not maintain backwards compatibility with
    previous protocol versions. If a plugin author wishes, they may branch their plugin program’s behaviour
    to be compatible with multiple versions of the protocol, or multiple major versions of sigstore and cosign.
* arg 2
  * JSON of initialization options and method arguments.
* stdout
  * JSON of method return values.

See [./common/interface.go](./common/interface.go) for the full JSON schema.

### Implementation

> Plugin authors must implement the `kms.SignerVerifier` interface methods in their chosen language. Each method will inbvoke your program once, and the reponse will be parsed from stdout. 

Author: fixed

  * for auto-invoking the correct interface methods, given an implementation. stdin will be the message argument for the methods that sign and verify data. The plugin author may do post-processing with the returned PluginResp.
* `WriteResponse(wr io.Writer, resp *PluginResp) error`
  * , should the author wish to construct and return their own PluginResp and exit.
* `WriteErrorResponse(wr io.Writer, err error) error`
  * , should the author simply send an error message and exit.

## Development

Changes to the `SignerVerifier` interface are to be handled in [./signer.go's](./signer.go) `PluginClient` and [./handler/dispatch.go's](./handler/dispatch.go) `Dispatch()`.

### Adding New Methods or Method Options

Adding new methods or options are *not* necessarily breaking changes to the schemas, so we may consider these to be minor version increments, both to the protocol version and the sigstore version.

### Removing Methods

Removing methods, or altering their signatures will break the schemas and will require major version increments, both to the protocl version and the sigstore version.

### Example Plugin

> We have an example plugin in [test/cliplugin/localkms](../../../.././test/cliplugin/localkms).

Author: fixed.

## Design

We follow [kubectl's style](https://kubernetes.io/docs/tasks/extend-kubectl/kubectl-plugins/#writing-kubectl-plugins) of plugins. Any language that can create CLI programs can be used to make a plugin.

### Usage

Plugins are separate programs on your system's PATH, named in the scheme `sigstore-kms-[name]`, like `sigstore-kms-my-hsm`. They can be invoked with cosign like `cosign [sub-command] --key "my-hsm://my-key-id" ...
`

### Protocol

The main program will invoke the program with these specifications:

* stdin
  * Data to be signed or verified.
* arg 1
  * A number identifying the version of this protocol.
    * In the future we may change this protocol, either the encoding or the formatting of argument and return values.
    Protocol changes will result in a major-version bump for the library.
    When changing the protocol, new versions of sigstore will not maintain backwards compatibility with
    previous protocol versions. If a plugin author wishes, they may branch their plugin program’s behaviour
    to be compatible with multiple versions of the protocol, or multiple major versions of sigstore and cosign.
* arg 2

>   * JSON of initialization options and method arguments.

Reviewer 2: why could these not be combined?

Author: It's possible we use a very different protocol than the current version. In order to keep things simple for future plugin programs, I think we should keep this first argument simple to parse.

* stdout
  * JSON of method return values.

See [./common/interface.go](./common/interface.go) for the full JSON schema.

> The plugin program must first exit before sigstore begins parsing responses.

Reviewer 2: why is that? In general I dont think you could enforce that

Author: Currently, we invoke the plugin program with `Cmd.Output()`, which waits until the plugin program exits, because:
1. We need to make sure that the main program received all of the intended output from the plugin program.
2. For the sake of users we may want to guarantee that plugin programs do eventually exit.

If we don't wait for the plugin program to exit, the protocol will need to be slightly more complex: read a full line of stdout before parsing the `PluginResp`. The newline character serves as a delimeter.

#### Error Handling

The plugin program’s stderr will be redirected to the main program’s stderr. This way, the main program may also see the plugin program’s debug messages.

Plugin authors may return errors with `PluginResp.ErrorMessage`, but the plugin's exit status will be ignored.

### Implementation

> Plugin authors must implement the `kms.SignerVerifier` interface methods in their chosen language. Each method will invoke your program once, and the reponse will be parsed from stdout.

Reviewer 2: ```suggestion
Plugin authors must implement the `kms.SignerVerifier` interface methods in their chosen language. Each method will invoke your program once, and the response will be parsed from stdout.
```

also it needs to logically implement the interface, not strictly (since you're not passing anything Go-specific over the wire from cosign/other tools?

Author: fixed

Exit status is ignored. Your program's stderr will be redirected to the main program, and errors you wish to return must be serialized in `PluginResp.ErrorMessage` in stdout.

For authors using Go, we vend some helper functions to help you get started. See [handler](./handler/README.md)

## Development

Changes to the `SignerVerifier` interface are to be handled in [./signer.go's](./signer.go) `PluginClient` and [./handler/dispatch.go's](./handler/dispatch.go) `Dispatch()`.

### Adding New Methods or Method Options

Adding new methods or options are *not* necessarily breaking changes to the schemas, so we may consider these to be minor version increments, both to the protocol version and the sigstore version.

### Removing Methods

> Removing methods, or altering their signatures will break the schemas and will require major version increments, both to the protocl version and the sigstore version.

Reviewer 2: ```suggestion
Removing methods, or altering their signatures will break the schemas and will require major version increments, both to the protocol version and the sigstore version.
```

Author: fixed
